Update instructions for container execution on windows and macos with Singularity and Docker

nandan75 opened the pull request:
Mergining changes to Nandan_changes branch. Will delete rest

I have not tested docker on Pawsey as yet. Will try and do it today and update readme accordingly

nandan75 committed:
Merged changes to Readme.md from other branches and deleted the additional branches

nandan75 committed:
Merge branch 'main' into Nandan_changes

georgiesamaha commented:
@nandan75  can you rename the PR and provide details in the description/comments here so I (as your reviewer) can understand the context for changes you've made?

nandan75 commented:
Hi @georgiesamaha 

@natbutter and I discussed/ debugged the execution of the RStudio container on multiple platforms

 The general consensus was -
- Suggest using docker if running the container on a local machine
   - We both tested the docker commands on multiple machines (All mac)
   -  The docker command ran fine for me (and @natbutter) on non-M1 macs
   - It did not run to completion on M1 mac for me (and @hlydecker - we are almost sure that it is M1 for Henry)
   - Please test docker  on a windows machine when you feel free.

- Singularity command runs fine on Pawsey Nimbus
- I have not tested Docker on pawsey Nimbus as yet - I will try it this afternoon if possible, else after I am back

- The overall structure of the Readme.md page is also modified, a bit.

nandan75 committed:
Removed a small typo at dcoker run command

nandan75 committed:
Finalised steps in Readme.md for specific operating systems 

Completed testing the container using docker on Pawsey Nimbus. Ran fine

Finalised writing specific steps for running the container in following ways 

1) Docker
    - Local mac machine
    - Pawsey Nimbus
2) Singularity 
    - Pawsey Nimbus

nandan75 committed:
Update README.md

Removed a small typo around $RSERVER_PASSWORD

nandan75 committed:
Update README.md

Small text updates

georgiesamaha commented:
Can run container following the mac instructions on Windows machine. **Caveat: I used Mobaxterm which is a (linux) terminal emulator.**

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
+- If you are on your local machine, we recommend that you run this container using Docker.  
+- We have tested running this container using Docker on local mac machines and also on a Pawsey Nimbus instance.
+  
+### On a local (mac) machine
```
Split this out into 2 subsections: 

```
### On a local machine 

#### MacOS 

#### Windows

georgiesamaha replied:
For windows you will need to suggest users run docker using a linux terminal emulator like MobaXterm

nandan75 replied:
Relavant comments added

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
+### On a local (mac) machine
 To run this container:
-
 - Start docker
```
How do I do that?

nandan75 replied:
Start docker by `double-clicking` on the `Docker` app icon in the `Applications` menu.

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
 To run this container:
-
 - Start docker
 - Mount a directory housing your data (specify which paths you need to mount), and run the RStudio server instance using the following command: 
```
How do I do that?

nandan75 replied:
Actually this is part of the RStudio server command .. I have now rephrased the sentence and explained the mounting after the command

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
 ## Run with Docker
-
+- If you are on your local machine, we recommend that you run this container using Docker.  
+- We have tested running this container using Docker on local mac machines and also on a Pawsey Nimbus instance.
```
Whats the relevant specs of a pawsey nimbus instance?

nandan75 replied:
We have tested this container using Docker on local mac machines and also on the Pawsey Nimbus cloud instance `Pawsey Bio - Ubuntu 22.04 - 2023-06`

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
 
-- Open up a browser window at http://localhost:8787/ 
-- Enter the username: rstudio
+You can now navigate a local web browser at http://localhost:8787/
```
Change this to: 

Open a web browser and enter http://localhost:8787/ in the address bar

nandan75 replied:
Changed

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
+You can now navigate a local web browser at http://localhost:8787/
+
+- Enter the username: `rstudio`
+- Enter the password you gave it e.g `yourpassword` as seen in the above command
```
Just fyi, when i tried it with yourpassword it doesnt work. You'll need to remind them to set a password of their own.

nandan75 replied:
I think the issue could have been more because I had not put yourpassword in single quotes as 'yourpassword'. Me and Nate had seen this before. The command is now changed.

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
+- Enter the username: `rstudio`
+- Enter the password you gave it e.g `yourpassword` as seen in the above command
+
+### On Pawsey Nimbus instance
```
Change this to:

```
### Pawsey Nimbus cloud
```

nandan75 replied:
DONE

georgiesamaha commented on README.md:
```diff
@@ -16,25 +16,39 @@ Or
+- Enter the password you gave it e.g `yourpassword` as seen in the above command
+
+### On Pawsey Nimbus instance
+Assuming you are running Docker on a headless remote server, you must ensure you connect to the host with port forwarding enabled if you wish to view the RStudio session on your local machine.
```
What does it mean to run docker on a headless remote server??? Simplify this for your audience.

nandan75 replied:
Assuming you are running Docker on a remote server without a graphical user interface (GUI)

changed

georgiesamaha commented on README.md:
```diff
@@ -63,9 +77,11 @@ sudo docker push sydneyinformaticshub/rnaseq-rstudio:4.1.0
-If you are on a machine with no Docker (like a HPC environment), you may wish to use [Singularity](https://docs.sylabs.io/guides/3.7/admin-guide/installation.html) as an alternative.
+If you are on a machine with no Docker (like a HPC environment or a Pawsey Nimbus instance), you may wish to use [Singularity](https://docs.sylabs.io/guides/3.7/admin-guide/installation.html) as an alternative.
+
+The following steps can assist in running the container on a Pawsey NImbus instance.
```
Remove this line. Keep formatting and style consistent with the rest of your README. Replace with: 

```
### On Pawsey Nimbus cloud
```

nandan75 replied:
DONE

georgiesamaha commented on README.md:
```diff
@@ -80,34 +96,35 @@ mkdir -p /tmp/rstudio-server
 
 Then launch the server similar to the docker command but with additional configuration options required:
```
I'd suggest you avoid potential confusion and not mention docker here. Remove this line.

nandan75 replied:
ok DONE

georgiesamaha commented on README.md:
````diff
@@ -80,34 +96,35 @@ mkdir -p /tmp/rstudio-server
+RSERVER_PASSWORD=$(openssl rand -base64 15)
+echo $RSERVER_PASSWORD
+```
+Please notedown this RSERVER_PASSWORD as you will need when you log on to RStudio from a browser.
````
Fix typo here and reword for clarity. I suggest: 

```
Take note of the password returned by your echo command. You will need this to log into RStudio in your browser. 
```

nandan75 replied:
DONE

nandan75 commented:
Do you want to include that option? 
- Something like a "Mobaxterm" is essential for a windows user to connect/perform minimal command line stuff anyways right?

georgiesamaha commented on README.md:
```diff
@@ -80,34 +96,35 @@ mkdir -p /tmp/rstudio-server
-Gene $RSERVER_PASSWORD
-
-
 PASSWORD=$RSERVER_PASSWORD singularity exec \
```
Preface this code block with: 

```
Launch the R server:
```

nandan75 replied:
ok

georgiesamaha commented on README.md:
````diff
@@ -80,34 +96,35 @@ mkdir -p /tmp/rstudio-server
-
 ```bash
-ssh -L 8787:localhost:8787 -i "your-ssh-key" ubuntu@123.456.789
+ssh -L 8787:localhost:8787 -i "your-ssh-key" ubuntu@146.118.69.202
````
Please dont provide your actual IP address here. Replace with 146.118.XX.XXX

nandan75 replied:
OK

georgiesamaha commented on README.md:
```diff
@@ -80,34 +96,35 @@ mkdir -p /tmp/rstudio-server
-Assuming you are running Singularity on a headless remote server, you must ensure you connect to the host with port forwarding enabled if you wish to view the RStudio session on your local machine.
+* `rserver --auth-none=0 --auth-pam-helper-path=pam-helper --server-user ubuntu` executes the command in the container, in this case "rserver" with various options. A key option here is the `--server-user rstudio` which assumes the user on the host machine is called **rstudio**, this is necessary with most Singularity setups to maintain the security mapping between users of the host and in the co ...
 
+Assuming you are running singularity on a headless remote server, you must ensure you connect to the host with port forwarding enabled if you wish to view the RStudio session on your local machine.
```
Same comment as per the docker command. Simplify for your audience and use less jargon where possible please.

georgiesamaha replied:
Also does this need to be in its own section? Why would I need to do this instead of just running the command above?

nandan75 replied:
I am not too sure, what you mean here! 

Can you please change this as appropriate.

georgiesamaha reviewed (changes requested):
Hi @nandan75, a few things: 
- I can run mac docker command with linux terminal emulator. 
- You'll need to include this in your README 
- Some suggestions regarding instructions in README that improve readability and accessibility for people who may not be as technically proficient as you.

nandan75 committed:
Applied changes as per review #10 by Georgie

nandan75 reviewed (commented):
I have addresses most (except 1 ) comments , made the changes and pushed them back to nandan_changes branch.

Cheers.

nandan75 committed:
Applied additonal changes as per review #10 by Georgie

nandan75 committed:
Changed some small typos

georgiesamaha committed:
Clean typos, fix singularity explainer, reformat
